tools/monitoring/mysql/alert.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	"database/sql"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/knative/test-infra/tools/monitoring/config"
+	"github.com/knative/test-infra/tools/monitoring/mail"
+)
+
+const (
+	alertInsertStmt = `
+		INSERT INTO Alerts (
+			Sent, ErrorPattern
+		) VALUES (?,?)`
+
+	alertUpdateStmt = `
+	UPDATE Alerts
+	SET Sent = (?)
+	WHERE ErrorPattern = (?)`
+
+	emailTemplate = `In the past %v, 
+The number of occurrences of the following error pattern reached threshold:
+%s
+
+Hint for diagnose & recovery: %s
+`
+)
+
+type MailConfig struct {
+	*mail.Config
+	recipients []string
+}
+
+func (m *MailConfig) sendAlert(errorPattern string, config *config.SelectedConfig) error {
+	log.Printf("sending alert...")
+	subject := fmt.Sprintf("Error pattern reached alerting threshold: %s", errorPattern)
+	body := fmt.Sprintf(emailTemplate, config.Duration(), errorPattern, config.Hint)
+
+	return m.Send(m.recipients, subject, body)
+}
+
+// Alert checks alert condition and alerts table and send alert mail conditionally
+func (m *MailConfig) Alert(errorPattern string, config *config.SelectedConfig, db *sql.DB) (bool, error) {
+	if ok, err := config.CheckAlertCondition(errorPattern, db); err != nil || !ok {
+		return false, err
+	}
+
+	queryTemplate, err := checkAlertsTable(errorPattern, config.Duration(), db)
+	if err != nil || queryTemplate == "" {
+		return false, err
+	}
+
+	if err := updateAlertsTable(queryTemplate, errorPattern, db); err != nil {
+		return false, err
+	}
+
+	err = m.sendAlert(errorPattern, config)
+	return err == nil, err
```

Comment:
Why return `err == nil`? For all cases where it returns err, it always return false. I think you can change the signature to `func (m *MailConfig) Alert(errorPattern string, config *config.SelectedConfig, db *sql.DB) error` and get rid of the pair-value return.

Reply:
the bool is returns whether an alert is sent. It can be false even when there is no error

Reply:
What is the bool value used for after it's returned from `Alert`? 

Also, I don't think it's a good idea to send the email from mysql package. I think the mail message generation can be moved to `mail` package and the actual sending email can be moved to the main app.

For the sql package, it can return the bool value on sending or not sending the alert.

Reply:
I think it is important to expose whether an alert is actually sent instead of keeping it within the function
I agree with your 2nd paragraph. In fact I don't think we need an mysql package at all, because we have sql calls in other packages as well and we already moved the common sql functionalities to the shared folder. Furthermore, I think we can move the mail package to shared folder as well, because none of the content their is coupled with monitoring
I prefer to get this checked in first before we do more refactoring. I have refactoring done locally, but if I upload it to this PR there will be a lot of change.